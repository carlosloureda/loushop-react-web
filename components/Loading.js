import styled from "styled-components";
import React from "react";

import PropTypes from "prop-types";

const LoadingStyles = styled.div`
  /* padding: 2rem;
  background: white;
  margin: 2rem 0;
  border: 1px solid rgba(0, 0, 0, 0.05);
  border-left: 5px solid red;
  p {
    margin: 0;
    font-weight: 100;
  }
  strong {
    margin-right: 1rem;
  } */
`;

const DisplayLoading = ({ loading }) => {
  if (!loading) return null;
  return (
    <LoadingStyles>
      <p data-test="graphql-loading">
        <strong>Loading ...</strong>
      </p>
    </LoadingStyles>
  );
};

DisplayLoading.defaultProps = {
  loading: false
};

DisplayLoading.propTypes = {
  loading: PropTypes.boolean
};

export default DisplayLoading;
