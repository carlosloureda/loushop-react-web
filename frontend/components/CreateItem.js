import React, { Component } from "react";
import { Mutation } from "react-apollo";
import gql from "graphql-tag";
import Router from "next/router";
import Form from "./styles/Form";
import formatMoney from "../lib/formatMoney";
import Error from "./ErrorMessage";

const CREATE_ITEM_MUTATION = gql`
  mutation CREATE_ITEM_MUTATION(
    $title: String!
    $description: String!
    $price: Int!
    $image: String
    $largeImage: String
  ) {
    createItem(
      title: $title
      description: $description
      price: $price
      image: $image
      largeImage: $largeImage
    ) {
      id
    }
  }
`;

class CreateItem extends Component {
  state = {
    item: {
      title: "",
      description: "",
      image: "",
      largeImage: "",
      price: ""
    },
    error: {}
  };

  handleChange = e => {
    const { name, type, value } = e.target;
    const val = val && type === "number" ? parseFloat(value) : value;
    this.setState({
      item: {
        ...this.state.item,
        [name]: val
      }
    });
  };
  //TODO: Lot of work with iamges, show loading when image is uploading
  uploadFile = async e => {
    console.log("upload file ...");
    const files = e.target.files;
    if (!e.target.files) return;
    const data = new FormData();
    data.append("file", files[0]);
    data.append("upload_preset", "loushop");
    const res = await fetch(
      "https://api.cloudinary.com/v1_1/carloslorueda/image/upload",
      {
        method: "POST",
        body: data
      }
    );
    const file = await res.json();
    //TODO: Error with maxium files, set waringin message on before loading
    if (file.error) {
      this.setState({ error: file.error });
      return;
    }
    this.setState({
      item: {
        ...this.state.item,
        image: file.secure_url,
        largeImage:
          file.eager && file.eager.length
            ? file.eager[0].secure_url
            : file.secure.url
      }
    });
  };
  render() {
    const { item } = this.state;
    return (
      <Mutation refetchQueries={} mutation={CREATE_ITEM_MUTATION} variables={item}>
        {/* {(createItem, payload) => { */}
        {/* {(createItem, {loading, error, called, data}) => { */}
        {(createItem, { loading, error }) => (
          <Form
            onSubmit={async e => {
              e.preventDefault();
              const res = await createItem();
              Router.push({
                pathname: "/item",
                query: { id: res.data.createItem.id }
              });
            }}
          >
            <Error error={error} />
            <Error error={this.state.error} />
            <fieldset disabled={loading} aria-busy={loading}>
              <label htmlFor="file">
                Image
                <input
                  type="file"
                  id="file"
                  name="file"
                  placeholder="Upoad an image"
                  required
                  onClick={() => this.setState({ error: {} })}
                  onChange={this.uploadFile}
                />
                {item.image && (
                  <img width="200" src={item.image} alt="Upload preview" />
                )}
              </label>
              <label htmlFor="title">
                Title
                <input
                  type="text"
                  id="title"
                  name="title"
                  placeholder="Title"
                  required
                  value={item.title}
                  onChange={this.handleChange}
                />
              </label>
              <label htmlFor="price">
                Price
                <input
                  type="number"
                  id="price"
                  name="price"
                  placeholder="Price"
                  required
                  value={item.price}
                  onChange={this.handleChange}
                />
              </label>
              <label htmlFor="description">
                Description
                <textarea
                  id="description"
                  name="description"
                  placeholder="Description"
                  required
                  value={item.description}
                  onChange={this.handleChange}
                />
              </label>
            </fieldset>
            <button type="submit">Submit</button>
          </Form>
        )}
      </Mutation>
    );
  }
}
export default CreateItem;
export { CREATE_ITEM_MUTATION };
