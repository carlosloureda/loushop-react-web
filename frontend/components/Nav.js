import Link from "next/link";
import NavStyles from "./styles/NavStyles";
import User from "./User";
import Signout from "./Signout";
// if (me) return <p>{me.name}</p>; // return null; // )}
const Nav = () => (
  <User>
    {({ data: { me }, loading, error }) => (
      <NavStyles>
        <Link href="/items">
          <a>Shop</a>
        </Link>
        {me && (
          <>
            <Link href="/sell">
              <a>Sell</a>
            </Link>
            <Link href="/orders">
              <a>orders</a>
            </Link>
            <Link href="/me">
              <a>account</a>
            </Link>
            <Signout />
          </>
        )}
        {!me && (
          <Link href="/signup">
            <a>Sign In</a>
          </Link>
        )}
      </NavStyles>
    )}
  </User>
);

export default Nav;
