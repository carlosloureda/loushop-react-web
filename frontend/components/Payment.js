import React, { Component } from "react";
import StripeCheckout from "react-stripe-checkout";
import calcTotalPrice from "../lib/calcTotalPrice";
import User, { CURRENT_USER_QUERY } from "./User";
import gql from "graphql-tag";
import { Mutation } from "react-apollo";

const CREATE_ORDER_MUTATION = gql`
  mutation createOrder($token: String!) {
    createOrder(token: $token) {
      id
      charge
      total
      items {
        id
        title
      }
    }
  }
`;

const totalItems = cart =>
  cart.reduce(
    (tally, cartItem) => tally + cartItem.item.price * cartItem.quantity,
    0
  );

const stripePublicKey = "pk_test_h4IkfdNKJXal2GPvNpu3Yzh3";

class Payment extends Component {
  onToken = async (res, createOrder) => {
    console.log("token is: ", res.id);
    const order = await createOrder({
      variables: { token: res.id }
    }).catch(e => alert(e.message));
  };
  render() {
    return (
      <User>
        {({ data: { me } }) => (
          <Mutation
            mutation={CREATE_ORDER_MUTATION}
            refetchQueries={[{ query: CURRENT_USER_QUERY }]}
          >
            {createOrder => (
              <StripeCheckout
                amount={calcTotalPrice(me.cart)}
                description={`Order of ${totalItems(me.cart)} items`}
                image={me.cart && me.cart.length ? me.cart[0].item.image : null}
                currency="USD"
                stripeKey={stripePublicKey}
                email={me.email}
                token={res => this.onToken(res, createOrder)}
              >
                {this.props.children}
              </StripeCheckout>
            )}
          </Mutation>
        )}
      </User>
    );
  }
}

export default Payment;
