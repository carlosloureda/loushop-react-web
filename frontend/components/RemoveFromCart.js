import React, { Component } from "react";
import { Mutation } from "react-apollo";
import styled from "styled-components";
import PropTypes from "prop-types";
import gql from "graphql-tag";
import { CURRENT_USER_QUERY } from "./User";

const REMOVE_FROM_CART_MUTATION = gql`
  mutation removeFromCart($id: ID!) {
    removeFromCart(id: $id) {
      id
    }
  }
`;

const BigButton = styled.button`
  font-size: 3rem;
  background: nnone;
  border: 0;
  &:hover {
    color: ${Props => Props.theme.red};
    cursor: pointer;
  }
`;

class RemoveFromCart extends Component {
  static propTypes = {
    id: PropTypes.string.isRequired
  };

  // This gets called as soon as we get a response back from tghe server after a
  // mutation has been performed. With the optimisticResponse gets called twice,
  //   one on client and another time when server responses
  update = (cache, payload) => {
    const data = cache.readQuery({ query: CURRENT_USER_QUERY });
    const cartItemId = payload.data.removeFromCart.id;
    data.me.cart = data.me.cart.filter(cartItem => cartItem.id !== cartItemId);
    cache.writeQuery({ query: CURRENT_USER_QUERY, data });
  };
  render() {
    return (
      <Mutation
        mutation={REMOVE_FROM_CART_MUTATION}
        variables={{ id: this.props.id }}
        refetchQueries={[{ query: CURRENT_USER_QUERY }]}
        update={this.update}
        optimisticResponse={{
          __typename: "Mutation",
          removeFromCart: {
            __typename: "CartItem",
            id: this.props.id
          }
        }}
      >
        {(removeFromCart, { loading, error }) => (
          <BigButton
            disabled={loading}
            onClick={() => {
              removeFromCart().catch(err => alert(err.message));
            }}
            title="Delete Item"
          >
            &times;
          </BigButton>
        )}
      </Mutation>
    );
  }
}
export default RemoveFromCart;
