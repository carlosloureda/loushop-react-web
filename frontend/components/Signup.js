import React, { Component } from "react";
import { Mutation } from "react-apollo";
import gql from "graphql-tag";
import Form from "./styles/Form";
import Error from "./ErrorMessage";
import { CURRENT_USER_QUERY } from "./User";

const SIGNUP_MUTATION = gql`
  mutation SIGNUP_MUTATION(
    $email: String!
    $name: String!
    $password: String!
  ) {
    signup(email: $email, name: $name, password: $password) {
      id
      email
      name
    }
  }
`;

class Signup extends Component {
  state = {
    name: "",
    email: "",
    password: ""
  };
  saveToState = e => {
    this.setState({ [e.target.name]: e.target.value });
  };
  render() {
    const { name, email, password } = this.state;
    return (
      <Mutation
        mutation={SIGNUP_MUTATION}
        variables={this.state}
        refetchQueries={[{ query: CURRENT_USER_QUERY }]}
      >
        {(signup, { error, loading }) => (
          <Form
            method="post"
            onSubmit={async e => {
              e.preventDefault();
              const a = await signup(); // Apollo gets the erorrs by himself
              this.setState({
                email: "",
                name: "",
                email: ""
              });
            }}
          >
            <fieldset disabled={loading} aria-busy={loading}>
              <h2>Sign up for an Account</h2>
              <Error error={error} />
              <label htmlFor="email">
                Email
                <input
                  type="text"
                  name="email"
                  placeholder="email"
                  value={email}
                  onChange={this.saveToState}
                />
              </label>
              <label htmlFor="name">
                Name
                <input
                  type="text"
                  name="name"
                  placeholder="name"
                  value={name}
                  onChange={this.saveToState}
                />
              </label>
              <label htmlFor="password">
                Password
                <input
                  type="papssword"
                  name="password"
                  placeholder="password"
                  value={password}
                  onChange={this.saveToState}
                />
              </label>
              <button type="submit">Sign up!</button>
            </fieldset>
          </Form>
        )}
      </Mutation>
    );
  }
}
export default Signup;
