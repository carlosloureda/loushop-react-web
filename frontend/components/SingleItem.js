import React, { Component } from "react";
import PropTypes from "prop-types";
import gql from "graphql-tag";
import { Query } from "react-apollo";
import styled from "styled-components";
import Error from "./ErrorMessage";
import Head from "./Meta";

const SingleItemStyles = styled.div`
  max-width: 1200px;
  margin: 2rem auto;
  box-shadow: ${props => props.theme.bs};
  display: grid;
  grid-auto-columns: 1fr;
  grid-auto-flow: column;
  min-height: 800px;
  img {
    width: 100%;
    height: 100%;
    object-fit: contain;
  }
  .details {
    margin: 3rem;
    font-size: 2rem;
  }
`;

const SINGLE_ITEM_QUERY = gql`
  query SINGLE_ITEM_QUERY($id: ID!) {
    item(where: { id: $id }) {
      id
      title
      price
      description
      largeImage
    }
  }
`;
class SingleItem extends Component {
  render() {
    return (
      <Query query={SINGLE_ITEM_QUERY} variables={{ id: this.props.id }}>
        {({ error, loading, data }) => {
          if (error) return <Error error={error} />;
          else if (loading) return <p>Loading...</p>;
          if (!data.item) return <p>No Item Found for {this.props.id}</p>;
          const item = data.item;
          return (
            <SingleItemStyles>
              {/* Side effects with NExt. it overrides the head */}
              <Head>
                <title>LouShop | {item.title}</title>
              </Head>
              {/* Single Item Component!!! {this.props.id} */}
              <img src={item.largeImage} alt={item.title} />
              <div className="details">
                <h2>Viewing {item.title}</h2>
                <p>{item.description}</p>
              </div>
            </SingleItemStyles>
          );
        }}
      </Query>
    );
  }
}

SingleItem.propTypes = {
  id: PropTypes.string.isRequired
};

export default SingleItem;
export { SINGLE_ITEM_QUERY };
