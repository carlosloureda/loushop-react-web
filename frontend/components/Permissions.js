import { Query } from "react-apollo";
import Error from "./ErrorMessage";
import Loading from "./Loading";
import gql from "graphql-tag";
import Table from "./styles/Table";
import SickButton from "./styles/SickButton";
import PropTypes from "prop-types";

const possiblePermissions = [
  "ADMIN",
  "USER",
  "ITEMCREATE",
  "ITEMUPDATE",
  "ITEMDELETE",
  "PERMISSIONUPDATE"
];
const ALL_USERS_QUERY = gql`
  query {
    users {
      id
      name
      email
      permissions
    }
  }
`;
const Permissions = props => (
  <Query query={ALL_USERS_QUERY}>
    {({ data, loading, error }) => (
      <div>
        <Error error={error} />
        {loading && <Loading loading={loading} />}
        <div>
          <h2>Manage Permissions</h2>
          <Table>
            <thead>
              <tr>
                <th>Name</th>
                <th>Email</th>
                {possiblePermissions.map(permission => (
                  <th key={permission}>{permission}</th>
                ))}
                <th>Action</th>
              </tr>
            </thead>
            <tbody>
              {data &&
                data.users.map(user => <User user={user} key={user.id} />)}
            </tbody>
          </Table>
        </div>
      </div>
    )}
  </Query>
);

class User extends React.Component {
  static propTypes = {
    user: PropTypes.shape({
      name: PropTypes.string,
      email: PropTypes.string,
      id: PropTypes.string,
      permissions: PropTypes.array
    }).isRequired
  };
  state = {
    // not a good thing to set state with props , in this case yeah!
    permissions: this.props.user.permissions
  };
  handlePermissionChange = e => {
    const checked = e.target;
    let updatedPermissions = [...this.state.permissions]; //copy
    // need to add the new permission
    if (checked.checked) {
      updatedPermissions.push(checked.value);
    } else {
      //remove
      updatedPermissions = updatedPermissions.filter(
        permission => permission !== checked.value
      );
    }
    this.setState({
      permissions: updatedPermissions
    });
  };
  render() {
    const { user } = this.props;
    return (
      <tr>
        <td>{user.name}</td>
        <td>{user.email}</td>
        {possiblePermissions.map(permission => (
          <td key={permission}>
            <label htmlFor={`${user.id}-permission-${permission}`}>
              <input
                id={`${user.id}-permission-${permission}`}
                type="checkbox"
                checked={this.state.permissions.includes(permission)}
                onChange={this.handlePermissionChange}
                value={permission}
              />
            </label>
          </td>
        ))}
        <td>
          <SickButton>Update</SickButton>
        </td>
      </tr>
    );
  }
}

export default Permissions;
