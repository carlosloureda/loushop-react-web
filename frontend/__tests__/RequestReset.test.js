import { mount } from "enzyme";
import wait from "waait";
import { MockedProvider } from "react-apollo/test-utils";
import RequestReset, {
  REQUEST_RESET_MUTATION
} from "../components/RequestReset";
import { CURRENT_USER_QUERY } from "../components/User";
import toJSON from "enzyme-to-json";
import { fakeUser, fakeCartItem } from "../lib/testUtils";

const mocks = [
  {
    request: {
      query: REQUEST_RESET_MUTATION,
      variables: { email: "carloslouredaparrado@gmail.com" }
    },
    result: {
      data: { requestReset: { message: "success", __typename: "Message" } }
    }
  },
  {
    request: { query: CURRENT_USER_QUERY },
    result: {
      data: {
        me: {
          ...fakeUser(),
          cart: [fakeCartItem(), fakeCartItem(), fakeCartItem()]
        }
      }
    }
  }
];

// const signedInMocksWithCartItems = [
//   {
//     request: { query: CURRENT_USER_QUERY },
//     result: {
//       data: {
//         me: {
//           ...fakeUser(),
//           cart: [fakeCartItem(), fakeCartItem(), fakeCartItem()]
//         }
//       }
//     }
//   }
// ];

describe("<RequestRest />", () => {
  it("renders and matches snapshot", async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <RequestReset />
      </MockedProvider>
    );
    const form = wrapper.find('form[data-test="form"]');
    expect(toJSON(form)).toMatchSnapshot();
  });

  it("calls the mutation", async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <RequestReset />
      </MockedProvider>
    );
    // simulate typing an email
    wrapper.find("input").simulate("change", {
      target: { name: "email", value: "carloslouredaparrado@gmail.com" }
    });
    // submit the form
    wrapper.find("form").simulate("submit");
    await wait();
    wrapper.update();
    expect(wrapper.find("p").text()).toContain(
      "Success! Check your email for a reset link!"
    );
  });
});
