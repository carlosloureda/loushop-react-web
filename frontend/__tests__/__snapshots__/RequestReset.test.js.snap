// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RequestRest /> renders and matches snapshot 1`] = `null`;
